### PR TITLE
Move kubeproxy ComponentConfig external types to `k8s.io/kube-proxy`

### PR DESCRIPTION
**Pull request by Lion-Wei**
**What this PR does / why we need it**:
This PR implements most of kubernetes/community#2354 for the kube-proxy.
The PR:
- Moves k8s.io/kubernetes/pkg/proxy/apis/kubeproxyconfig as-is to k8s.io/kubernetes/pkg/proxy/apis/config as agreed
- Moves the external types to the new staging repo k8s.io/kube-proxy, in the k8s.io/kube-proxy/config/v1beta1 package.
- Makes k8s.io/kubernetes/pkg/proxy/apis/config/v1beta1 source the types from k8s.io/kube-proxy/config/v1beta1. The defaulting and conversion code is kept in this package as before.
- All references to these packages have been updated.

Ref #67233

**Special notes for your reviewer**:

**Release note**:
```release-note
kube-proxy v1beta1 external ComponentConfig types are now available in the `k8s.io/kube-proxy` repo
```

### CODE REVIEW

File: pkg/proxy/apis/kubeproxyconfig/v1alpha1/register.go [hunk 1]
```diff
@@ -30,21 +30,13 @@ var SchemeGroupVersion = schema.GroupVersion{Group: GroupName, Version: "v1alpha
 var (
 	// TODO: move SchemeBuilder with zz_generated.deepcopy.go to k8s.io/api.
```
Comment (luxas) [inline pkg/proxy/apis/kubeproxyconfig/v1alpha1/register.go]: remove these comments here

File: staging/src/k8s.io/kube-proxy/config/v1alpha1/register.go [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+)
+
+// GroupName is the group name use in this package
+const GroupName = "kubeproxy.config.k8s.io"
+
+// SchemeGroupVersion is group version used to register these objects
+var SchemeGroupVersion = schema.GroupVersion{Group: GroupName, Version: "v1alpha1"}
+
+var (
+	// TODO: move SchemeBuilder with zz_generated.deepcopy.go to k8s.io/api.
```
Comment (luxas) [inline staging/src/k8s.io/kube-proxy/config/v1alpha1/register.go]: remove these comments, only misleading

File: pkg/proxy/apis/config/doc.go [hunk 3]
```diff
@@ -16,4 +16,4 @@ limitations under the License.
 
 // +k8s:deepcopy-gen=package
 
-package kubeproxyconfig // import "k8s.io/kubernetes/pkg/proxy/apis/kubeproxyconfig"
+package config // import "k8s.io/kubernetes/pkg/proxy/apis/config"
```
Comment (luxas) [inline pkg/proxy/apis/config/doc.go]: this should probably be done already in the first commit, as well as moving OWNERS and BUILD I think

File: vendor/k8s.io/kube-proxy [hunk 4]
```diff
@@ -0,0 +1 @@
+../../staging/src/k8s.io/kube-proxy/
```
Comment (luxas) [inline vendor/k8s.io/kube-proxy]: do this already in the first commit

**Review by luxas — COMMENTED** [review]
Thanks for this PR, great work @Lion-Wei :tada: :clap:!!

Quick comment on the commit structure I'd like to see:
 - Move `pkg/proxy/apis/kubeproxyconfig` to `pkg/proxy/apis/config`. All files should be moved and package names should be updated
 - Rename all references in the rest of the codebase to `pkg/proxy/apis/config` instead
 - Start referencing `ClientConnectionConfiguration` from `k8s.io/apimachinery/pkg/apis/config` instead of duplicating it in the package
 - Add `k8s.io/kube-proxy/config`, move external types there from `pkg/proxy/apis/config/v1alpha1`.
 - All autogenerated stuff in the last commit

Alternatively, break out the two first commits in a separate PR (move `pkg/proxy/apis/{kubeproxy,}config`), and the third commit (move `ClientConnectionConfiguration` reference) to a separate PR as well. So then we'd get two PRs first that are more scoped and hence easier to merge.

File: pkg/proxy/apis/config/v1alpha1/defaults.go [hunk 5]
```diff
@@ -1,5 +1,5 @@
 /*
-Copyright 2015 The Kubernetes Authors.
+Copyright 2018 The Kubernetes Authors.
```
Comment (xiangpengzhao) [inline pkg/proxy/apis/config/v1alpha1/defaults.go]: Moving and changing the existing files don't necessarily change the year here, IMO. @bgrant0607 WDYT?

File: pkg/proxy/apis/kubeproxyconfig/v1alpha1/doc.go [hunk 6]
```diff
@@ -15,8 +15,9 @@ limitations under the License.
...
 // +k8s:deepcopy-gen=package
-// +k8s:conversion-gen=k8s.io/kubernetes/pkg/proxy/apis/kubeproxyconfig
-// +k8s:openapi-gen=true
+// +k8s:conversion-gen=k8s.io/kubernetes/pkg/proxy/apis/config
+// +k8s:conversion-gen-external-types=k8s.io/kube-proxy/config/v1alpha1
 // +k8s:defaulter-gen=TypeMeta
+// +k8s:defaulter-gen-input=../../../../../vendor/k8s.io/kube-proxy/config/v1alpha1
 
-package v1alpha1 // import "k8s.io/kubernetes/pkg/proxy/apis/kubeproxyconfig/v1alpha1"
+package v1alpha1 // import "k8s.io/kubernetes/pkg/proxy/apis/config/v1alpha1"
```
Comment (sttts) [inline pkg/proxy/apis/kubeproxyconfig/v1alpha1/doc.go]: please add the GroupName tag

File: staging/src/k8s.io/kube-proxy/config/v1alpha1/doc.go [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// +k8s:deepcopy-gen=package
+
+package v1alpha1 // import "k8s.io/kube-proxy/config/v1alpha1"
```
Comment (sttts) [inline staging/src/k8s.io/kube-proxy/config/v1alpha1/doc.go]: please add the GroupName tag

**Comment by sttts** [comment]
Sgtm. Some nit.

**Comment by sttts** [comment]
If the staging dir is new, don't forget the meta files like COPYRIGHT, OWNER, etc.

**Comment by luxas** [comment]
Additionally, the golint ignore list needs to be updated, import-boss files needs to be updated with the new location to whitelist, and a Godeps file needs to be created

**Comment by Lion-Wei** [comment]
@sttts @luxas Thanks for reviewing.
I reorganized the commit for the sake of reviewing, and all previous comment are addressed. Please take another look, thanks.

File: pkg/proxy/apis/config/v1alpha1/types.go [hunk 8]
```diff
@@ -17,32 +17,17 @@ limitations under the License.
...
-}
-
 // KubeProxyIPTablesConfiguration contains iptables-related configuration
 // details for the Kubernetes proxy server.
 type KubeProxyIPTablesConfiguration struct {
 	// masqueradeBit is the bit of the iptables fwmark space to use for SNAT if using
 	// the pure iptables proxy mode. Values must be within the range [0, 31].
 	MasqueradeBit *int32 `json:"masqueradeBit"`
-	// masqueradeAll tells kube-proxy to SNAT everything if using the pure iptables proxy mode.
+	// masqueradeAll tells kube-proxy to SNAT everything if using the pure iptables proxy mode.cmd/kubeadm/app/apis/kubeadm/v1alpha3/conversion.go
```
Comment (luxas) [inline pkg/proxy/apis/config/v1alpha1/types.go]: this seems like a small copy-paste error

File: pkg/proxy/apis/config/v1alpha1/doc.go [hunk 9]
```diff
@@ -16,7 +16,8 @@ limitations under the License.
 
 // +k8s:deepcopy-gen=package
 // +k8s:conversion-gen=k8s.io/kubernetes/pkg/proxy/apis/config
-// +k8s:openapi-gen=true
+// +k8s:conversion-gen-external-types=k8s.io/kube-proxy/config/v1alpha1
 // +k8s:defaulter-gen=TypeMeta
-
+// +k8s:defaulter-gen-input=../../../../../vendor/k8s.io/kube-proxy/config/v1alpha1
+// +groupName=networking.k8s.io
```
Comment (luxas) [inline pkg/proxy/apis/config/v1alpha1/doc.go]: this is not true I think, should be `kubeproxy.config.k8s.io` I'm pretty sure

File: pkg/proxy/apis/config/v1alpha1/register.go [hunk 10]
```diff
@@ -30,21 +30,13 @@ var SchemeGroupVersion = schema.GroupVersion{Group: GroupName, Version: "v1alpha
 var (
 	// TODO: move SchemeBuilder with zz_generated.deepcopy.go to k8s.io/api.
```
Comment (luxas) [inline pkg/proxy/apis/config/v1alpha1/register.go]: this comment can/should be removed

File: staging/src/k8s.io/kube-proxy/config/v1alpha1/doc.go [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// +k8s:deepcopy-gen=package
+// +groupName=networking.k8s.io
```
Comment (luxas) [inline staging/src/k8s.io/kube-proxy/config/v1alpha1/doc.go]: should be `kubeproxy.config.k8s.io`

File: staging/src/k8s.io/kube-proxy/config/v1alpha1/register.go (hunk 2 above)
Comment (luxas) [reply staging/src/k8s.io/kube-proxy/config/v1alpha1/register.go]: ping pls remove

**Review by luxas — APPROVED** [review]
/approve
Thanks a lot @Lion-Wei :+1: !!
This LGTM, except for the `networking.k8s.io` groupname tag error. Please fix then this is ok by me

**Pull request merged**